src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,2059 @@
+        super().__init__()
+        # 1st residual block
+        self.self_attn = NgramMultiheadAttention(config)
+        self.self_attn_layer_norm = LayerNorm(config.hidden_size)
+
+        # 2nd residual block
+        if config.add_cross_attention:
+            self.cross_attn = SelfAttention(config, config.num_decoder_attention_heads)
+            self.cross_attn_layer_norm = LayerNorm(config.hidden_size)
+
+        # 3rd residual block
+        self.feed_forward = FeedForwardBlock(config, config.decoder_ffn_dim)
+        self.feed_forward_layer_norm = LayerNorm(config.hidden_size)
+
+    def forward(
+        self,
+        hidden_states,
+        encoder_hidden_states=None,
+        encoder_attn_mask=None,
+        layer_state=None,
+        attention_mask=None,
+        output_attentions=False,
+        extended_predict_attention_mask=None,
+        main_relative_position_buckets=None,
+        predict_relative_position_buckets=None,
+        position_ids=None,
+    ):
+        layer_state = layer_state if layer_state is not None else {}
+
+        # 1st residual block
+        ngram_attention_output, self_attn_weights, self_attn_weights_ngram = self.self_attn(
+            hidden_states=hidden_states,
+            layer_state=layer_state,
+            need_weights=False,
+            attention_mask=attention_mask,
+            extended_predict_attention_mask=extended_predict_attention_mask,
+            main_relative_position_buckets=main_relative_position_buckets,
+            predict_relative_position_buckets=predict_relative_position_buckets,
+            position_ids=position_ids,
+        )
+        hidden_states = self.self_attn_layer_norm(hidden_states + ngram_attention_output)
+
+        cross_attn_weights = None
+        if encoder_hidden_states is not None:
+            # 2nd residual block
+            attention_output, cross_attn_weights = self.cross_attn(
+                hidden_states=hidden_states,
+                key_value_states=encoder_hidden_states,
+                attention_mask=encoder_attn_mask,
+                layer_state=layer_state,  # mutates layer state
+            )
+            hidden_states = self.cross_attn_layer_norm(attention_output + hidden_states)
+
+        # 3rd residual block
+        feed_forward_output = self.feed_forward(hidden_states)
+        hidden_states = self.feed_forward_layer_norm(feed_forward_output + hidden_states)
+
+        return (
+            hidden_states,
+            self_attn_weights,
+            self_attn_weights_ngram,
+            cross_attn_weights,
+            layer_state,
+        )  # just self_attn weights for now, following t5, layer_state = cache for decoding
+
+
+@add_start_docstrings(
+    "The standalone encoder part of the ProphetNetModel.",
+    PROPHETNET_START_DOCSTRING,
+)
+class ProphetNetEncoder(ProphetNetPreTrainedModel):
+    def __init__(self, config: ProphetNetConfig, word_embeddings: nn.Embedding = None):
```

Comment:
The `word_embeddings` parameter is not documented in `PROPHETNET_START_DOCSTRING` so its documentation should be added here.